src/net_processing.cpp

```diff
@@ -2839,15 +2846,19 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             // we have a chain with at least nMinimumChainWork), and we ignore
             // compact blocks with less work than our tip, it is safe to treat
             // reconstructed compact blocks as having been requested.
-            ProcessNewBlock(chainparams, pblock, /*fForceProcessing=*/true, &fNewBlock);
-            if (fNewBlock) {
+            CValidationState dos_state;
+            ProcessNewBlock(chainparams, pblock, dos_state, /*fForceProcessing=*/true, &fNewBlock);
+            if (fNewBlock && dos_state.IsValid()) {
                 pfrom->nLastBlockTime = GetTime();
             } else {
+                if (!dos_state.IsValid()) {
+                    BlockChecked(*pblock, dos_state, connman);
+                }
                 LOCK(cs_main);
                 mapBlockSource.erase(pblock->GetHash());
             }
-            LOCK(cs_main); // hold cs_main for CBlockIndex::IsValid()
-            if (pindex->IsValid(BLOCK_VALID_TRANSACTIONS)) {
+            if (dos_state.IsValid()) {
```

Comment:
This seems like a behaviour change. If `ConnectBlock()` fails for this block, then `pindex->nStatus` will be updated with `BLOCK_FAILED_VALID`. Previously, that would prevent this branch from being executed, but now we're only looking at the return `CValidationState` from `CheckBlock()` and `AcceptBlock()`.

If this is an intentional behaviour change, can you pull it out into a separate commit and comment on why in the commit log?

Reply:
Don't think it matters either way (though also don't think it was *intentional* per se). It's not so much an important change as it is an accidental tweak thats probably a bit better but shouldn't matter. Once we get the block we have to mark it as received, but we cannot do that if the block is malleated. If the block is invalid, we don't care too much - essentially anyone who sends us the block is gonna get the banhammer (if the block is not received via compact blocks).

Reply:
If it doesn't matter either way, can you leave it unchanged? Trying to pick through what exactly `MarkBlockAsReceived()` is doing and whether changing whether it's called or not is adding quite a lot of burden to review.

Reply:
IMO, this is much cleaner, even if the behavior proper doesn't matter. a) reaching into validation to figure out what happened after calling PNB is...a massive layer violation, this should be returned, not figured out by side-effects, b) it means one chunk of cs_main less (even if its only in the uncommon case), and c) it matters a lot for #16323.

Reply:
ok, if it does matter, can you separate into its own commit, with a commit log saying _why_ it matters a lot :)

Reply:
Two different definitions of "matter", I suppose. Anyway, I pulled it out on its own.

Reply:
Tests that clarify and assert the points in this discussion would be reassuring.

Reply:
There really is no material difference in behavior here. Its much *cleaner* code, but the behavior is the same. Hence the two different definitions of "matter".